### ISSUE TITLE
Enhancement: Support for returning `AsyncGenerator[ServerSentEventMessage, None]` at route handler

### ISSUE THREAD

**Issue by winstxnhdw** [litestar-org/litestar#3851]
### Summary

Currently, Litestar forces you to create an intermediary function that yields `ServerSentEventMessage` then wrap the generator in `ServerSentEvent`.

```python
class SomeController(Controller):
    async def intermediary_function(self) -> AsyncGenerator[ServerSentEventMessage, None]:
        yield ServerSentEventMessage(data='Doing this..')
        yield ServerSentEventMessage(data='Doing that..')
        yield ServerSentEventMessage(data='Finished doing this and that..')

    @get()
    async def actual_server_sent_event_handler(self) -> ServerSentEvent:
        return ServerSentEvent(self.intermediary_function())   
```

This is not an ideal pattern as you now have a method that isn't a route handler in your controller. If you were to abstract it out as a function, you'll now have to find an elegant place to define this function. In any case, this is an annoying problem.

### Basic Example

Ideally, this is what you'd want. It has good code locality and you don't have to jump through functions or files to understand the full context of the code.

```python
class SomeController(Controller):
    @get()
    async def actual_server_sent_event_handler(self) -> AsyncGenerator[ServerSentEventMessage, None]:
        yield ServerSentEventMessage(data='Doing this..')
        yield ServerSentEventMessage(data='Doing that..')
        yield ServerSentEventMessage(data='Finished doing this and that..')
```

### Drawbacks and Impact

I have a pseudo solution to this using decorators that has zero runtime cost aside from the initial injection at startup. Ideally, the actual solution should not have to depend on any decorators. 

```python
from collections.abc import AsyncIterator, Awaitable, Callable, Iterator
from functools import partial, update_wrapper
from inspect import isasyncgenfunction

from litestar.response import ServerSentEvent, ServerSentEventMessage


async def async_handler[**P](
    handler: Callable[P, Iterator[ServerSentEventMessage] | AsyncIterator[ServerSentEventMessage]],
    *args: P.args,
    **kwargs: P.kwargs,
) -> ServerSentEvent:
    return ServerSentEvent(handler(*args, **kwargs))


def server_sent_event[**P](
    handler: Callable[P, Iterator[ServerSentEventMessage] | AsyncIterator[ServerSentEventMessage]],
) -> Callable[P, ServerSentEvent | Awaitable[ServerSentEvent]]:
    new_handler: Callable[P, ServerSentEvent | Awaitable[ServerSentEvent]] = (
        partial(async_handler, handler)
        if isasyncgenfunction(handler)
        else lambda *args, **kwargs: ServerSentEvent(handler(*args, **kwargs))
    )

    new_handler = update_wrapper(new_handler, handler)
    new_handler.__annotations__['return'] = ServerSentEvent

    return new_handler


class SomeController(Controller):
    @get()
    @server_sent_event
    async def actual_server_sent_event_handler(self) -> AsyncGenerator[ServerSentEventMessage, None]:
        yield ServerSentEventMessage(data='Doing this..')
        yield ServerSentEventMessage(data='Doing that..')
        yield ServerSentEventMessage(data='Finished doing this and that..')
```

With this solution, I am still able to correctly generate OpenAPI docs and have no issues with DI. I am quite confident that this can be implemented without issues. If you know of any potential drawbacks, I'd like to hear it.

---
> [!NOTE]  
> While we are open for sponsoring on [GitHub Sponsors](https://github.com/sponsors/litestar-org/) and 
> [OpenCollective](https://opencollective.com/litestar), we also utilize [Polar.sh](https://polar.sh/) to engage in pledge-based sponsorship.
>
> Check out all issues funded or available for funding [on our Polar.sh dashboard](https://polar.sh/litestar-org)
> * If you would like to see an issue prioritized, make a pledge towards it!
> * We receive the pledge once the issue is completed & verified
> * This, along with engagement in the community, helps us know which features are a priority to our users.

<a href="https://polar.sh/litestar-org/litestar/issues/3851">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/litestar-org/litestar/issues/3851/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/litestar-org/litestar/issues/3851/pledge.svg">
</picture>
</a>

**Comment by provinzkraut** [comment]
There is actually a major drawback, or rather footgun, which is why this feature is (currently) not implemented:

The injected dependencies are scoped to the route handler, not the generator. 

The effect of this is that if you inject a dependency with a cleanup step directly into the generator function, it might already be closed before you start iterating, since the route handler has already returned.

Right now, this is (at least somewhat) obvious, since you request them in the handler, with them being available in the generator only if you pass them in / capture them with a closure. This is in-line with standard generator behaviour in Python, e.g. 

```python
def consume_thing(thing: Thing) -> Generator[str, None, None]:
  yield thing.get_value()
  yield thing.get_value()


def thing_handler() -> Generator[str, None, None]: 
  with Thing() as thing:
   # Thing will be closed once this function returns, not when the generator finishes
   return consume_thing(thing)  
```

With your proposal, it would be even more counterintuitive why a dependency might already have run its cleanup step.

**Current implementation**

```python
async def provide_thing() -> AsyncGenerator[Thing, None]:
  async with Thing() as thing:
    yield thing

@get(dependencies{thing: provide_thing})
async def handler(thing: Thing) -> ServerSentEvent:
  async def intermediary_function() -> AsyncGenerator[ServerSentEventMessage, None]:
    yield ServerSentEventMessage(data=thing.get_value())
  
  # provide_thing's cleanup step will be called right after this line
  return ServerSentEvent(intermediary_function())  
```

**Proposed implementation**
```python
async def provide_thing() -> AsyncGenerator[Thing, None]:
  async with Thing() as thing:
    yield thing

@get(dependencies{thing: provide_thing})
async def handler(thing: Thing) -> AsyncGenerator[ServerSentEventMessage, None]:
  # provide_thing's cleanup step will be called right before this line executes
  yield ServerSentEventMessage(data=thing.get_value())
```

**How to do this properly**

```python
async def provide_thing() -> AsyncGenerator[Thing, None]:
  async with Thing() as thing:
    yield thing

@get()
async def handler() -> ServerSentEvent:
  async def intermediary_function() -> AsyncGenerator[ServerSentEventMessage, None]:
    async with Thing() as thing:
      yield ServerSentEventMessage(data=thing.get_value())
  
  return ServerSentEvent(intermediary_function())  
```

The current behaviour

**Comment by provinzkraut** [comment]
A section in the documentation about this would be good though.

**Comment by winstxnhdw** [comment]
Your suggestion doesn’t solve the fact that dependencies would clean-up prematurely if they’re used within the generator.

**Comment by provinzkraut** [comment]
> Your suggestion doesn’t solve the fact that dependencies would clean-up prematurely if they’re used within the generator.
> 
> Your proposal has zero benefits to DX nor does it help alleviate the counter intuitiveness of DI in generator streams.

I'm not proposing anything. The proposal mentioned in my comment is *your* proposal

**Comment by winstxnhdw** [comment]
I meant your “how to do this properly”. At least with my proposal, you could emit a warning to not use DI when using async generators. Litestar relies so much on tribal knowledge and have no plans to address any of it. Are we on maintenance mode?

**Comment by provinzkraut** [comment]
> Litestar relies so much on tribal knowledge

I don't see this as "tribal knowledge", as it's the same behaviour you'd get in plain Python when mixing context managers and generators, as I've shown in my first example ([another issue where I'm talking about this for reference](https://github.com/litestar-org/litestar/issues/3842)). The counter intuitiveness you're talking about would apply to that as well. 

I am not disagreeing that this behaviuor *can* be non-obvious if you're new to generators and their mechanics, and it's not an uncommon pitfall, but I don't think it's a good idea to introduce something in Litestar specifically to address a common pitfall in *Python*.

While this may help you personally in your understanding, I believe that it would leave more people confused who are aware of these mechanics, and to whom an explanation of "this works just like a context manager + generator" (because it is just that really :slightly_smiling_face:) is sufficient. And if that explanation is *not* sufficient, what we should do instead is e.g. point to the stdlib docs / article that explains the mechanics behind it. Which is why I said a section in our documentation about this would be good. 

PRs are open, so you're very much invited to add something that you think would explain this in a way that avoids the issues you're seeing :slightly_smiling_face: 

> and have no plans to address any of it. Are we on maintenance mode?

Can you elaborate?

**Comment by provinzkraut** [comment]
> Again, this is only expected behaviour if you’ve seen the Litestar source code. Unless that is what you expect of your users? 

It is explained in the documentation: https://docs.litestar.dev/latest/usage/dependency-injection.html#dependencies-with-yield-cleanup-step.

> The cleanup stage is executed after the handler function returns
 
And dependencies with a cleanup step being generators is not an implementation detail either: It's how you define them.

You are correct in saying that this behaviour exists because Litestar uses generators to implement dependencies with a cleanup step, but the fact that it does is not hidden. And since they are used, I don't think it's unreasonable to assume that they'd behave just like they normally would. In fact it's deviating from this that would be unexpected. 

> The point of this issue is to discuss the approach before making the PR but it seems you don’t want to do that either, seeing as you’ve closed this issue immediately.

This issue was a feature request, which was not accepted, due to the reasons stated in my original comment. I suggested that if you feel like this part of Litestar's behaviour is unexpected, you may open a PR to improve the documentation.

**Comment by winstxnhdw** [comment]
I am not following. If Litestar just uses generators to implement the clean-up step, DI wouldn’t properly clean up regardless of using a generator in the route handler or not.

See the following.

```python

class FieldManager:
    def __init__(self):
        self.field = None

    def __enter__(self):
        # Initialize the field
        self.field = ["hello"]
        return self.field

    def __exit__(self, exc_type, exc_value, traceback):
        # Print statement to confirm cleanup
        print("FieldManager: Cleaning up by emptying the field list.")
        self.field.clear()  # Empty the list instead of deleting it

# Generator function that uses the FieldManager context manager
def generate_with_field():
    try:
        with FieldManager() as field:
            yield field  # Yield the field with the "hello" list
    finally:
        # Cleanup happens here after the generator is fully consumed
        print("generate_with_field: Generator cleanup triggered.")

# Route handler function that takes a generator and yields from it
def stream_route_handler(a):
    yield a

# Use the generator function and pass it to the route handler
a = next(generate_with_field())
route_handler = stream_route_handler(a)

# Print the outputs from the route handler
for i in range(3):  # Limit to 3 iterations for demonstration
    print(next(route_handler))
```

`a` would still prematurely clean up.

**Comment by provinzkraut** [comment]
> `a` would still prematurely clean up.

That is because you are not saving a reference to the generator, but only the first value it yields. Again, this is not something Litestar specific, but just general generator semantics.

You example is a bit contrived, so I simplified to demonstrate:

**This behaves as expected**
```python
def generate_with_field():
    try:
        yield "hello"
    finally:
        print("cleanup in generator")

def handler_function(some_dependency):
    print(f"received value {some_dependency!r}")
    print("exiting route handler")

a = generate_with_field()
print("calling handler")
handler_function(next(a))
```
and prints:
```
calling handler
received value 'hello'
exiting route handler
cleanup in generator
```

**This cleans up prematurely**
```python
def generate_with_field():
    try:
        yield "hello"
    finally:
        print("cleanup in generator")

def handler_function(some_dependency):
    print(f"received value {some_dependency!r}")
    print("exiting route handler")

print("calling handler")
handler_function(next(generate_with_field()))
```
and prints:
```
calling handler
cleanup in generator
received value 'hello'
exiting route handler
```

The reason this is happening is because by not saving a reference to the generator, it will get garbage collected immediately, which will in turn trigger the `finally` clause.

Notice the difference between
```python
stream_route_handler(next(generate_with_field()))
``` 
and
```python
a = generate_with_field()
stream_route_handler(next(a))
```

You can read about this here: https://docs.python.org/3/reference/expressions.html#yieldexpr

<hr>

Litestar does not rely on the garbage collector to clean up the generators, but instead does this explicitly. The result in this example would still be the same:

**What Litestar is actually doing**



```python
def generate_with_field():
    try:
        yield "hello"
    finally:
        print("cleanup in generator")

def stream_route_handler(some_dependency):
    print(f"received value {some_dependency!r}")
    print("exiting route handler")

print("calling handler")
a = generate_with_field()
stream_route_handler(next(a))
print("calling cleanup")

next(a, None)
```

this prints: 
```text
calling handler
received value 'hello'
exiting route handler
calling cleanup
cleanup in generator
```

**Comment by winstxnhdw** [comment]
Okay, so I copied what you did and saved the reference to the generator instead and as you can see, the premature clean up step would not happen.

```python
class FieldManager:
    def __init__(self):
        self.field = None

    def __enter__(self):
        # Initialize the field
        self.field = ["hello"]
        return self.field

    def __exit__(self, exc_type, exc_value, traceback):
        # Print statement to confirm cleanup
        print("FieldManager: Cleaning up by emptying the field list.")
        self.field.clear()  # Empty the list instead of deleting it

def generate_with_field():
    try:
        with FieldManager() as field:
            yield field  # Yield the field with the "hello" list
    finally:
        # Cleanup happens here after the generator is fully consumed
        print("generate_with_field: Generator cleanup triggered.")

# Route handler function that takes a value and yields it in a loop
def stream_route_handler(a):
    while True:
        yield a  # Yield the passed value

# Usage Example
if __name__ == "__main__":
    # Create the generator
    a = generate_with_field()

    # Pass the result of `next(a)` into the route handler
    route_handler = stream_route_handler(next(a))
    for i in range(3):  # Limit to 3 iterations for demonstration
        print(next(route_handler))
```

**Comment by provinzkraut** [comment]
> Okay, so I copied what you did and saved the reference to the generator instead and as you can see, the premature clean up step still would not happen so why should users expect the clean up to happen in a server sent event?

That is because your example does not represent what is happening in Litestar, which is what I have illustrated in my first comment.

In fact, it showcases exactly the type of misunderstanding I was talking about, since your example is in line with how you assume your proposed feature could work, when in fact, it would not work that way. You are explicitly *not* doing the explicit cleanup step on the generator created by `generate_with_field` *after* the route handler function has returned, which is what Litestar does.

Change 

```python
route_handler = stream_route_handler(next(a))
```
to

```python
route_handler = stream_route_handler(next(a))
next(a)
```
And the output will be

```text
FieldManager: Cleaning up by emptying the field list.
generate_with_field: Generator cleanup triggered.
```

**The dependencies are scoped to the route handler function. They will be cleaned up after the handler function returns. It does not matter whether or not it returns a generator.**

**Comment by winstxnhdw** [comment]
Yes and that is the point I have been trying to push since the beginning. That this issue is in fact, Litestar specific.

**Comment by provinzkraut** [comment]
You are mixing up two things here:

1. Something that is Litestar specific: Dependencies are always scoped to the route handlers
2. Something that is not Litestar specific: How generators work

The perceived issues arises because of the mix between these two, but, as long as you are aware of 1) - which we explicitly call out in the documentation -, the behaviour is quite expected and, as I (and you :slightly_smiling_face:) demonstrated, easily reproducible in plain Python.

And the "issue" of dependencies being scoped to the route handler isn't so much of an issue as it's a design choice. I've explained to you here in the beginning [and in this other issue you've opened](https://github.com/litestar-org/litestar/issues/3842) why we won't change that for one specific type of handler function.

The whole "how to generators work?" topic only came up because you implied that understanding the reasoning would require "tribal knowledge", whereas I maintain that it is not, and have tried explaining to you this point of view.

> It is hidden from the user. 

It is not hidden from the user. It is stated explicitly in the documentation that this is how it works.

> You are manually calling the next iteration. This is not something Python has forced onto you. You could have done the clean up at some arbitrary point later.

No, it is not. But as I have explain in various comments now, this is not something we will change for one type of handler function:

- https://github.com/litestar-org/litestar/issues/3842#issuecomment-2452938388
- https://github.com/litestar-org/litestar/issues/3851#issuecomment-2495929729

I'm not saying there's not a better way to handle dependencies for handler functions that return async generators, but making them behave differently implicitly isn't how we should go about this, as it would be confusing, and would also be a breaking change. [Here](https://github.com/litestar-org/litestar/issues/3842#issuecomment-2452938388) I have also shown you how this would possibly be fixed in version 3.0.

**Comment by provinzkraut** [comment]
I will be locking the conversation here now since I don't feel it to be productive anymore.
You are still welcome to open a PR with documentation changes you feel would address the tribal knowledge required :slightly_smiling_face:
